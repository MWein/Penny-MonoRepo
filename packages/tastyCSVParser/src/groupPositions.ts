import { TastyHistory } from './historyCSVToJson'
import { uniq } from 'lodash'
import { spreadHistoryByQuantity } from './utils/spreadByQuantity'
import { generateSymbol } from '@penny/test-helpers'

type TastyState = 'Open' | 'Closed'

export type TastyLeg = {
  state: TastyState,
  openDate: Date,
  closeDate: Date | null,
  symbol: string,
  side: 'Long' | 'Short',
  expiration: string,
  strike: number,
  premium: number,
  fees: number,
  realizedProfitLoss: number,
}

export type TastyPosition = {
  state: TastyState,
  underlying: string,
  openDate: Date,
  closeDate: Date | null,
  totalPremium: number,
  totalFees: number,
  realizedProfitLoss: number,
  legs: TastyLeg[]
}


const groupLegsIntoPositions = (underlying: string, legGroups: TastyLeg[][]): TastyPosition[] => {
  return legGroups.map(legs => {
    const isClosed = legs.every(leg => leg.state === 'Closed')
    const openDate = legs.map(leg => leg.openDate).sort()[0]
    const closeDate = isClosed ? legs.map(leg => leg.closeDate).filter(date => date).sort().reverse()[0] : null

    return {
      state: isClosed ? 'Closed' : 'Open',
      underlying,
      openDate,
      closeDate,
      totalPremium: legs.reduce((acc, leg) => acc + leg.premium, 0),
      totalFees: legs.reduce((acc, leg) => acc + leg.fees, 0),
      realizedProfitLoss: legs.reduce((acc, leg) => acc + leg.realizedProfitLoss, 0),
      legs
    }
  })
}


const groupPositionsForUnderlying = (underlying, history: TastyHistory[]) => {
  const underlyingHistory = spreadHistoryByQuantity(
    history
    .filter(history => history.underlying === underlying)
    .reverse()
  )

  const positions: TastyLeg[][] = []
  let currentLegs: TastyLeg[] = []


  underlyingHistory.forEach(history => {
    // @ts-ignore
    const symbol = generateSymbol(history.underlying, history.optionType.toLowerCase(), history.expiration, history.strike)

    if (history.action === 'Open') {
      currentLegs.push({
        state: 'Open',
        symbol,
        openDate: history.date,
        closeDate: null,
        side: history.side === 'Buy' ? 'Long' : 'Short',
        expiration: history.expiration,
        strike: history.strike,
        premium: history.amount,
        fees: history.fees,
        realizedProfitLoss: 0,
      })
    }
    
    if (history.action === 'Close') {
      const index = currentLegs.findIndex(leg => leg.symbol === symbol && leg.state === 'Open')
      const editLeg = currentLegs[index]
      currentLegs[index] = {
        ...editLeg,
        state: 'Closed',
        closeDate: history.date,
        realizedProfitLoss: editLeg.premium + history.amount
      }
    }

    // Move to closed if all current legs are closed
    if (currentLegs.every(leg => leg.state === 'Closed')) {
      positions.push(currentLegs)
      currentLegs = []
    }
  })

  if (currentLegs.length > 0) {
    positions.push(currentLegs)
  }
  return positions
}



export const groupPositions = (history: TastyHistory[]) => uniq(
  history
    .map(hist => hist.underlying))
    .flatMap(symbol => groupLegsIntoPositions(symbol, groupPositionsForUnderlying(symbol, history))
)
