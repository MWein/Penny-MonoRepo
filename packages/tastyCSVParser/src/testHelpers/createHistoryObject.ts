import { TastyHistory } from '../historyCSVToJson'

export const createHistoryObj = (
  date: Date,
  underlying: string,
  side: 'Buy' | 'Sell',
  action: 'Open' | 'Close',
  quantity: number,
  expiration: string,
  strike: number,
  optionType: 'Call' | 'Put',
  price: number,
  fees: number,
  amount: number,
): TastyHistory => ({
  date, underlying, side, action, quantity, expiration, strike, optionType, price, fees, amount,
})