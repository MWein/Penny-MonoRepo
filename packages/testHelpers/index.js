"use strict";
// These functions generate mock orders and position objects for use in automated tests
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generatePositionObject = exports.generateOrderObject = exports.generateSymbol = void 0;
var generateSymbol = function (symbol, type) {
    switch (type) {
        case 'stock':
            return symbol;
        case 'call':
            return "".concat(symbol, "1234C3214");
        case 'put':
            return "".concat(symbol, "1234P3214");
    }
};
exports.generateSymbol = generateSymbol;
var generateOrderObject = function (symbol, quantity, type, side, status, id) {
    if (quantity === void 0) { quantity = 1; }
    if (type === void 0) { type = 'stock'; }
    if (side === void 0) { side = 'sell_to_open'; }
    if (status === void 0) { status = 'pending'; }
    if (id === void 0) { id = 123456; }
    var ordClass = type === 'call' || type === 'put' ? 'option' : 'equity';
    var orderObj = {
        id: id,
        type: 'market',
        symbol: symbol,
        side: side,
        quantity: quantity,
        status: status,
        duration: 'pre',
        avg_fill_price: 0.00000000,
        exec_quantity: 0.00000000,
        last_fill_price: 0.00000000,
        last_fill_quantity: 0.00000000,
        remaining_quantity: 0.00000000,
        create_date: '2018-06-06T20:16:17.342Z',
        transaction_date: '2018-06-06T20:16:17.357Z',
        class: ordClass,
        //option_symbol: 'AAPL180720C00274000'
    };
    if (ordClass === 'option') {
        return __assign(__assign({}, orderObj), { option_symbol: (0, exports.generateSymbol)(symbol, type) });
    }
    return orderObj;
};
exports.generateOrderObject = generateOrderObject;
var generatePositionObject = function (symbol, quantity, type, cost_basis, date_acquired, id) {
    if (quantity === void 0) { quantity = 1; }
    if (type === void 0) { type = 'stock'; }
    if (cost_basis === void 0) { cost_basis = 100; }
    if (date_acquired === void 0) { date_acquired = '2019-01-31T17:05'; }
    if (id === void 0) { id = 123456; }
    return ({
        cost_basis: cost_basis,
        date_acquired: date_acquired,
        id: id,
        quantity: quantity,
        symbol: (0, exports.generateSymbol)(symbol, type)
    });
};
exports.generatePositionObject = generatePositionObject;
