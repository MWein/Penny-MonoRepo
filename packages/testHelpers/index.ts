// These functions generate mock orders and position objects for use in automated tests

type EquityType = 'stock' | 'call' | 'put'
type OrderSide = 'sell_to_open' | 'sell_to_close' | 'buy_to_open' | 'buy_to_close'
type OrderClass = 'option' | 'equity'

type MockOrderObject = {
  id: number,
  type: string,
  symbol: string,
  side: OrderSide,
  quantity: number,
  status: string,
  duration: string,
  avg_fill_price: number,
  exec_quantity: number,
  last_fill_price: number,
  last_fill_quantity: number,
  remaining_quantity: number,
  create_date: string,
  transaction_date: string,
  class: OrderClass,
  option_symbol?: string,
}

type MockPositionObject = {
  cost_basis: number,
  date_acquired: string,
  id: number,
  quantity: number,
  symbol: string,
}



export const generateSymbol = (
  symbol: string,
  type: EquityType,
  customExp: string | null = null,
  customStrike: number | null = null
) => {
  const expText = customExp ? customExp.replace(/-/g, '').slice(2) : '1234'

  let strikeText = '3214'
  if (customStrike) {
    const beforeDecimalDigits = `${customStrike}`.split('.')[0].split('')
    const afterDecimalDigits = `${customStrike}`.split('.')[1]?.split('') || []

    while (beforeDecimalDigits.length !== 5) {
      beforeDecimalDigits.unshift('0')
    }

    while (afterDecimalDigits.length !== 3) {
      afterDecimalDigits.push('0')
    }

    strikeText = beforeDecimalDigits.join('') + afterDecimalDigits.join('')
  }

  switch (type) {
  case 'stock':
    return symbol
  case 'call':
    return `${symbol}${expText}C${strikeText}`
  case 'put':
    return `${symbol}${expText}P${strikeText}`
  }
}


export const generateOrderObject = (
    symbol: string,
    quantity: number = 1,
    type: EquityType = 'stock',
    side: OrderSide = 'sell_to_open',
    status: string ='pending',
    id: number = 123456
  ) : MockOrderObject => {
  const ordClass: OrderClass = type === 'call' || type === 'put' ? 'option' : 'equity'

  const orderObj = {
    id,
    type: 'market',
    symbol,
    side,
    quantity,
    status,
    duration: 'pre',
    avg_fill_price: 0.00000000,
    exec_quantity: 0.00000000,
    last_fill_price: 0.00000000,
    last_fill_quantity: 0.00000000,
    remaining_quantity: 0.00000000,
    create_date: '2018-06-06T20:16:17.342Z',
    transaction_date: '2018-06-06T20:16:17.357Z',
    class: ordClass,
    //option_symbol: 'AAPL180720C00274000'
  }

  if (ordClass === 'option') {
    return {
      ...orderObj,
      option_symbol: generateSymbol(symbol, type)
    }
  }

  return orderObj
}


export const generatePositionObject = (
    symbol: string,
    quantity: number = 1,
    type: EquityType ='stock',
    cost_basis: number = 100,
    date_acquired: string = '2019-01-31T17:05',
    id: number = 123456
  ) : MockPositionObject =>
  ({
    cost_basis,
    date_acquired,
    id,
    quantity,
    symbol: generateSymbol(symbol, type)
  })