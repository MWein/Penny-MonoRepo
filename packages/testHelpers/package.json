{
  "name": "@penny/test-helpers",
  "version": "1.0.0",
  "description": "Package with test helpers for mocks",
  "scripts": {
    "test": "jest --coverage"
  }
}
