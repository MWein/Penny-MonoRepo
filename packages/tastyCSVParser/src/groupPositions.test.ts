import { groupPositions } from './groupPositions'
import { createHistoryObj } from './testHelpers/createHistoryObject'


describe('groupPositions', () => {
  it('Groups single position (all transactions on the same date)', () => {
    const history = [
      createHistoryObj(new Date('2022-01-01'), 'AAPL', 'Sell', 'Open', 1, '2022-01-01', 180, 'Call', 0.45, 1.00, 45),
      createHistoryObj(new Date('2022-01-01'), 'AAPL', 'Sell', 'Open', 1, '2022-01-01', 170, 'Put', 0.45, 1.00, 45),
    ]
    expect(groupPositions(history)).toMatchSnapshot()
  })


  it('Groups single position with 1 bought and sold leg', () => {
    const history = [
      createHistoryObj(new Date('2022-01-02'), 'AAPL', 'Buy', 'Close', 1, '2022-01-01', 180, 'Call', 0.12, 1.00, -12),
      createHistoryObj(new Date('2022-01-01'), 'AAPL', 'Sell', 'Open', 1, '2022-01-01', 180, 'Call', 0.45, 1.00, 45),
    ]
    expect(groupPositions(history)).toMatchSnapshot()
  })


  it('Groups multiple positions (different underlyings)', () => {
    const history = [
      createHistoryObj(new Date('2022-01-02'), 'AAPL', 'Buy', 'Close', 1, '2022-01-01', 180, 'Call', 0.12, 1.00, -12),
      createHistoryObj(new Date('2022-01-02'), 'TSLA', 'Buy', 'Close', 1, '2022-01-01', 180, 'Call', 0.12, 1.00, -12),
      createHistoryObj(new Date('2022-01-01'), 'AAPL', 'Sell', 'Open', 1, '2022-01-01', 180, 'Call', 0.45, 1.00, 45),
      createHistoryObj(new Date('2022-01-01'), 'TSLA', 'Sell', 'Open', 1, '2022-01-01', 180, 'Call', 0.45, 1.00, 45),
    ]
    expect(groupPositions(history)).toMatchSnapshot()
  })


  it('Shows a position as open if legs are open', () => {
    const history = [
      createHistoryObj(new Date('2022-01-01'), 'AAPL', 'Sell', 'Open', 1, '2022-01-01', 180, 'Call', 0.45, 1.00, 45),
    ]
    expect(groupPositions(history)[0].state).toEqual('Open')
    expect(groupPositions(history)).toMatchSnapshot()
  })


  it('Leg shows as long if bought to open', () => {
    const history = [
      createHistoryObj(new Date('2022-01-01'), 'AAPL', 'Buy', 'Open', 1, '2022-01-01', 180, 'Call', 0.45, 1.00, -45),
    ]
    expect(groupPositions(history)[0].legs[0].side).toEqual('Long')
    expect(groupPositions(history)).toMatchSnapshot()
  })


  it('Leg shows as short if sold to open', () => {
    const history = [
      createHistoryObj(new Date('2022-01-01'), 'AAPL', 'Sell', 'Open', 1, '2022-01-01', 180, 'Call', 0.45, 1.00, 45),
    ]
    expect(groupPositions(history)[0].legs[0].side).toEqual('Short')
    expect(groupPositions(history)).toMatchSnapshot()
  })


  it('Shows a position as open if legs are open but some legs have been closed', () => {
    const history = [
      createHistoryObj(new Date('2022-01-02'), 'AAPL', 'Buy', 'Close', 1, '2022-01-01', 180, 'Call', 0.12, 1.00, -12),
      createHistoryObj(new Date('2022-01-01'), 'AAPL', 'Sell', 'Open', 1, '2022-01-01', 180, 'Call', 0.45, 1.00, 45),
      createHistoryObj(new Date('2022-01-01'), 'AAPL', 'Sell', 'Open', 1, '2022-01-01', 180, 'Call', 0.45, 1.00, 45),
    ]
    expect(groupPositions(history)[0].state).toEqual('Open')
    expect(groupPositions(history)).toMatchSnapshot()
  })


  it('Shows a position as closed if all legs were closed out', () => {
    const history = [
      createHistoryObj(new Date('2022-01-02'), 'AAPL', 'Buy', 'Close', 1, '2022-01-01', 180, 'Call', 0.12, 1.00, -12),
      createHistoryObj(new Date('2022-01-02'), 'AAPL', 'Buy', 'Close', 1, '2022-01-01', 180, 'Call', 0.12, 1.00, -12),
      createHistoryObj(new Date('2022-01-01'), 'AAPL', 'Sell', 'Open', 1, '2022-01-01', 180, 'Call', 0.45, 1.00, 45),
      createHistoryObj(new Date('2022-01-01'), 'AAPL', 'Sell', 'Open', 1, '2022-01-01', 180, 'Call', 0.45, 1.00, 45),
    ]
    expect(groupPositions(history)[0].state).toEqual('Closed')
    expect(groupPositions(history)).toMatchSnapshot()
  })


  // // If a strangle was entered, and then one side was rolled, it would still be one position
  it('Groups date-overlapping history objects as a single positions', () => {
    const history = [
      createHistoryObj(new Date('2022-01-05'), 'AAPL', 'Sell', 'Open', 1, '2022-01-01', 150, 'Call', 0.35, 1.00, 35),
      createHistoryObj(new Date('2022-01-05'), 'AAPL', 'Buy', 'Close', 1, '2022-01-01', 180, 'Call', 0.25, 1.00, -25),
      createHistoryObj(new Date('2022-01-01'), 'AAPL', 'Sell', 'Open', 1, '2022-01-01', 180, 'Call', 0.45, 1.00, 45),
      createHistoryObj(new Date('2022-01-01'), 'AAPL', 'Sell', 'Open', 1, '2022-01-01', 120, 'Put', 0.45, 1.00, 45),
    ]
    const result = groupPositions(history)
    expect(result.length).toEqual(1)
    expect(result).toMatchSnapshot()
  })


  // // If a strangle was entered and closed in May, and another one opened in June, they are separate positions
  it('Groups non-date-overlapping history objects as separate positions', () => {
    const history = [
      createHistoryObj(new Date('2022-01-07'), 'AAPL', 'Sell', 'Open', 1, '2022-01-01', 180, 'Call', 0.45, 1.00, 45),
      createHistoryObj(new Date('2022-01-07'), 'AAPL', 'Sell', 'Open', 1, '2022-01-01', 120, 'Put', 0.45, 1.00, 45),

      createHistoryObj(new Date('2022-01-05'), 'AAPL', 'Buy', 'Close', 1, '2022-01-01', 180, 'Call', 0.25, 1.00, -25),
      createHistoryObj(new Date('2022-01-05'), 'AAPL', 'Buy', 'Close', 1, '2022-01-01', 120, 'Put', 0.15, 1.00, -15),
      createHistoryObj(new Date('2022-01-01'), 'AAPL', 'Sell', 'Open', 1, '2022-01-01', 180, 'Call', 0.45, 1.00, 45),
      createHistoryObj(new Date('2022-01-01'), 'AAPL', 'Sell', 'Open', 1, '2022-01-01', 120, 'Put', 0.45, 1.00, 45),
    ]
    const result = groupPositions(history)
    expect(result.length).toEqual(2)
    expect(result).toMatchSnapshot()
  })


  it('Groups positions with quantity greater than 1', () => {
    const history = [
      createHistoryObj(new Date('2022-01-07'), 'AAPL', 'Sell', 'Open', 1, '2022-01-01', 180, 'Call', 0.45, 1.00, 45),
      createHistoryObj(new Date('2022-01-07'), 'AAPL', 'Sell', 'Open', 1, '2022-01-01', 120, 'Put', 0.45, 1.00, 45),
      createHistoryObj(new Date('2022-01-05'), 'AAPL', 'Buy', 'Close', 1, '2022-01-01', 180, 'Call', 0.25, 1.00, -25),
      createHistoryObj(new Date('2022-01-05'), 'AAPL', 'Buy', 'Close', 1, '2022-01-01', 120, 'Put', 0.15, 1.00, -15),
      createHistoryObj(new Date('2022-01-01'), 'AAPL', 'Sell', 'Open', 1, '2022-01-01', 180, 'Call', 0.45, 1.00, 45),
      createHistoryObj(new Date('2022-01-01'), 'AAPL', 'Sell', 'Open', 2, '2022-01-01', 120, 'Put', 0.45, 1.00, 45),
    ]
    const result = groupPositions(history)
    expect(result).toMatchSnapshot()
  })


  it('Groups positions with quantity greater than 1, uniqs (since positions are spread by quantity in the function)', () => {
    const history = [
      createHistoryObj(new Date('2022-01-05'), 'AAPL', 'Buy', 'Close', 2, '2022-01-01', 180, 'Call', 0.25, 1.00, -25),
      createHistoryObj(new Date('2022-01-05'), 'AAPL', 'Buy', 'Close', 2, '2022-01-01', 120, 'Put', 0.15, 1.00, -15),
      createHistoryObj(new Date('2022-01-01'), 'AAPL', 'Sell', 'Open', 2, '2022-01-01', 180, 'Call', 0.45, 1.00, 45),
      createHistoryObj(new Date('2022-01-01'), 'AAPL', 'Sell', 'Open', 2, '2022-01-01', 120, 'Put', 0.45, 1.00, 45),
    ]
    const result = groupPositions(history)
    expect(result).toMatchSnapshot()
  })
})