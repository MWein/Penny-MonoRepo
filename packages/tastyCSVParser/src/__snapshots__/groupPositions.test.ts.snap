// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`groupPositions Groups date-overlapping history objects as a single positions 1`] = `
Array [
  Object {
    "closeDate": null,
    "legs": Array [
      Object {
        "closeDate": null,
        "expiration": "2022-01-01",
        "fees": 1,
        "openDate": 2022-01-01T00:00:00.000Z,
        "premium": 45,
        "quantity": 1,
        "realizedProfitLoss": 0,
        "side": "Short",
        "state": "Open",
        "strike": 120,
        "symbol": "AAPL220101P00120000",
      },
      Object {
        "closeDate": 2022-01-05T00:00:00.000Z,
        "expiration": "2022-01-01",
        "fees": 1,
        "openDate": 2022-01-01T00:00:00.000Z,
        "premium": 45,
        "quantity": 1,
        "realizedProfitLoss": 20,
        "side": "Short",
        "state": "Closed",
        "strike": 180,
        "symbol": "AAPL220101C00180000",
      },
      Object {
        "closeDate": null,
        "expiration": "2022-01-01",
        "fees": 1,
        "openDate": 2022-01-05T00:00:00.000Z,
        "premium": 35,
        "quantity": 1,
        "realizedProfitLoss": 0,
        "side": "Short",
        "state": "Open",
        "strike": 150,
        "symbol": "AAPL220101C00150000",
      },
    ],
    "openDate": 2022-01-01T00:00:00.000Z,
    "realizedProfitLoss": 20,
    "state": "Open",
    "totalFees": 3,
    "totalPremium": 125,
    "underlying": "AAPL",
  },
]
`;

exports[`groupPositions Groups multiple positions (different underlyings) 1`] = `
Array [
  Object {
    "closeDate": 2022-01-02T00:00:00.000Z,
    "legs": Array [
      Object {
        "closeDate": 2022-01-02T00:00:00.000Z,
        "expiration": "2022-01-01",
        "fees": 1,
        "openDate": 2022-01-01T00:00:00.000Z,
        "premium": 45,
        "quantity": 1,
        "realizedProfitLoss": 33,
        "side": "Short",
        "state": "Closed",
        "strike": 180,
        "symbol": "AAPL220101C00180000",
      },
    ],
    "openDate": 2022-01-01T00:00:00.000Z,
    "realizedProfitLoss": 33,
    "state": "Closed",
    "totalFees": 1,
    "totalPremium": 45,
    "underlying": "AAPL",
  },
  Object {
    "closeDate": 2022-01-02T00:00:00.000Z,
    "legs": Array [
      Object {
        "closeDate": 2022-01-02T00:00:00.000Z,
        "expiration": "2022-01-01",
        "fees": 1,
        "openDate": 2022-01-01T00:00:00.000Z,
        "premium": 45,
        "quantity": 1,
        "realizedProfitLoss": 33,
        "side": "Short",
        "state": "Closed",
        "strike": 180,
        "symbol": "TSLA220101C00180000",
      },
    ],
    "openDate": 2022-01-01T00:00:00.000Z,
    "realizedProfitLoss": 33,
    "state": "Closed",
    "totalFees": 1,
    "totalPremium": 45,
    "underlying": "TSLA",
  },
]
`;

exports[`groupPositions Groups non-date-overlapping history objects as separate positions 1`] = `
Array [
  Object {
    "closeDate": 2022-01-05T00:00:00.000Z,
    "legs": Array [
      Object {
        "closeDate": 2022-01-05T00:00:00.000Z,
        "expiration": "2022-01-01",
        "fees": 1,
        "openDate": 2022-01-01T00:00:00.000Z,
        "premium": 45,
        "quantity": 1,
        "realizedProfitLoss": 30,
        "side": "Short",
        "state": "Closed",
        "strike": 120,
        "symbol": "AAPL220101P00120000",
      },
      Object {
        "closeDate": 2022-01-05T00:00:00.000Z,
        "expiration": "2022-01-01",
        "fees": 1,
        "openDate": 2022-01-01T00:00:00.000Z,
        "premium": 45,
        "quantity": 1,
        "realizedProfitLoss": 20,
        "side": "Short",
        "state": "Closed",
        "strike": 180,
        "symbol": "AAPL220101C00180000",
      },
    ],
    "openDate": 2022-01-01T00:00:00.000Z,
    "realizedProfitLoss": 50,
    "state": "Closed",
    "totalFees": 2,
    "totalPremium": 90,
    "underlying": "AAPL",
  },
  Object {
    "closeDate": null,
    "legs": Array [
      Object {
        "closeDate": null,
        "expiration": "2022-01-01",
        "fees": 1,
        "openDate": 2022-01-07T00:00:00.000Z,
        "premium": 45,
        "quantity": 1,
        "realizedProfitLoss": 0,
        "side": "Short",
        "state": "Open",
        "strike": 120,
        "symbol": "AAPL220101P00120000",
      },
      Object {
        "closeDate": null,
        "expiration": "2022-01-01",
        "fees": 1,
        "openDate": 2022-01-07T00:00:00.000Z,
        "premium": 45,
        "quantity": 1,
        "realizedProfitLoss": 0,
        "side": "Short",
        "state": "Open",
        "strike": 180,
        "symbol": "AAPL220101C00180000",
      },
    ],
    "openDate": 2022-01-07T00:00:00.000Z,
    "realizedProfitLoss": 0,
    "state": "Open",
    "totalFees": 2,
    "totalPremium": 90,
    "underlying": "AAPL",
  },
]
`;

exports[`groupPositions Groups positions with quantity greater than 1 1`] = `
Array [
  Object {
    "closeDate": null,
    "legs": Array [
      Object {
        "closeDate": 2022-01-05T00:00:00.000Z,
        "expiration": "2022-01-01",
        "fees": 0.5,
        "openDate": 2022-01-01T00:00:00.000Z,
        "premium": 22.5,
        "quantity": 1,
        "realizedProfitLoss": 7.5,
        "side": "Short",
        "state": "Closed",
        "strike": 120,
        "symbol": "AAPL220101P00120000",
      },
      Object {
        "closeDate": null,
        "expiration": "2022-01-01",
        "fees": 0.5,
        "openDate": 2022-01-01T00:00:00.000Z,
        "premium": 22.5,
        "quantity": 1,
        "realizedProfitLoss": 0,
        "side": "Short",
        "state": "Open",
        "strike": 120,
        "symbol": "AAPL220101P00120000",
      },
      Object {
        "closeDate": 2022-01-05T00:00:00.000Z,
        "expiration": "2022-01-01",
        "fees": 1,
        "openDate": 2022-01-01T00:00:00.000Z,
        "premium": 45,
        "quantity": 1,
        "realizedProfitLoss": 20,
        "side": "Short",
        "state": "Closed",
        "strike": 180,
        "symbol": "AAPL220101C00180000",
      },
      Object {
        "closeDate": null,
        "expiration": "2022-01-01",
        "fees": 1,
        "openDate": 2022-01-07T00:00:00.000Z,
        "premium": 45,
        "quantity": 1,
        "realizedProfitLoss": 0,
        "side": "Short",
        "state": "Open",
        "strike": 120,
        "symbol": "AAPL220101P00120000",
      },
      Object {
        "closeDate": null,
        "expiration": "2022-01-01",
        "fees": 1,
        "openDate": 2022-01-07T00:00:00.000Z,
        "premium": 45,
        "quantity": 1,
        "realizedProfitLoss": 0,
        "side": "Short",
        "state": "Open",
        "strike": 180,
        "symbol": "AAPL220101C00180000",
      },
    ],
    "openDate": 2022-01-01T00:00:00.000Z,
    "realizedProfitLoss": 27.5,
    "state": "Open",
    "totalFees": 4,
    "totalPremium": 180,
    "underlying": "AAPL",
  },
]
`;

exports[`groupPositions Groups positions with quantity greater than 1, uniqs (since positions are spread by quantity in the function) 1`] = `
Array [
  Object {
    "closeDate": 2022-01-05T00:00:00.000Z,
    "legs": Array [
      Object {
        "closeDate": 2022-01-05T00:00:00.000Z,
        "expiration": "2022-01-01",
        "fees": 1,
        "openDate": 2022-01-01T00:00:00.000Z,
        "premium": 45,
        "quantity": 2,
        "realizedProfitLoss": 30,
        "side": "Short",
        "state": "Closed",
        "strike": 120,
        "symbol": "AAPL220101P00120000",
      },
      Object {
        "closeDate": 2022-01-05T00:00:00.000Z,
        "expiration": "2022-01-01",
        "fees": 1,
        "openDate": 2022-01-01T00:00:00.000Z,
        "premium": 45,
        "quantity": 2,
        "realizedProfitLoss": 20,
        "side": "Short",
        "state": "Closed",
        "strike": 180,
        "symbol": "AAPL220101C00180000",
      },
    ],
    "openDate": 2022-01-01T00:00:00.000Z,
    "realizedProfitLoss": 50,
    "state": "Closed",
    "totalFees": 2,
    "totalPremium": 90,
    "underlying": "AAPL",
  },
]
`;

exports[`groupPositions Groups single position (all transactions on the same date) 1`] = `
Array [
  Object {
    "closeDate": null,
    "legs": Array [
      Object {
        "closeDate": null,
        "expiration": "2022-01-01",
        "fees": 1,
        "openDate": 2022-01-01T00:00:00.000Z,
        "premium": 45,
        "quantity": 1,
        "realizedProfitLoss": 0,
        "side": "Short",
        "state": "Open",
        "strike": 170,
        "symbol": "AAPL220101P00170000",
      },
      Object {
        "closeDate": null,
        "expiration": "2022-01-01",
        "fees": 1,
        "openDate": 2022-01-01T00:00:00.000Z,
        "premium": 45,
        "quantity": 1,
        "realizedProfitLoss": 0,
        "side": "Short",
        "state": "Open",
        "strike": 180,
        "symbol": "AAPL220101C00180000",
      },
    ],
    "openDate": 2022-01-01T00:00:00.000Z,
    "realizedProfitLoss": 0,
    "state": "Open",
    "totalFees": 2,
    "totalPremium": 90,
    "underlying": "AAPL",
  },
]
`;

exports[`groupPositions Groups single position with 1 bought and sold leg 1`] = `
Array [
  Object {
    "closeDate": 2022-01-02T00:00:00.000Z,
    "legs": Array [
      Object {
        "closeDate": 2022-01-02T00:00:00.000Z,
        "expiration": "2022-01-01",
        "fees": 1,
        "openDate": 2022-01-01T00:00:00.000Z,
        "premium": 45,
        "quantity": 1,
        "realizedProfitLoss": 33,
        "side": "Short",
        "state": "Closed",
        "strike": 180,
        "symbol": "AAPL220101C00180000",
      },
    ],
    "openDate": 2022-01-01T00:00:00.000Z,
    "realizedProfitLoss": 33,
    "state": "Closed",
    "totalFees": 1,
    "totalPremium": 45,
    "underlying": "AAPL",
  },
]
`;

exports[`groupPositions Leg shows as long if bought to open 1`] = `
Array [
  Object {
    "closeDate": null,
    "legs": Array [
      Object {
        "closeDate": null,
        "expiration": "2022-01-01",
        "fees": 1,
        "openDate": 2022-01-01T00:00:00.000Z,
        "premium": -45,
        "quantity": 1,
        "realizedProfitLoss": 0,
        "side": "Long",
        "state": "Open",
        "strike": 180,
        "symbol": "AAPL220101C00180000",
      },
    ],
    "openDate": 2022-01-01T00:00:00.000Z,
    "realizedProfitLoss": 0,
    "state": "Open",
    "totalFees": 1,
    "totalPremium": -45,
    "underlying": "AAPL",
  },
]
`;

exports[`groupPositions Leg shows as short if sold to open 1`] = `
Array [
  Object {
    "closeDate": null,
    "legs": Array [
      Object {
        "closeDate": null,
        "expiration": "2022-01-01",
        "fees": 1,
        "openDate": 2022-01-01T00:00:00.000Z,
        "premium": 45,
        "quantity": 1,
        "realizedProfitLoss": 0,
        "side": "Short",
        "state": "Open",
        "strike": 180,
        "symbol": "AAPL220101C00180000",
      },
    ],
    "openDate": 2022-01-01T00:00:00.000Z,
    "realizedProfitLoss": 0,
    "state": "Open",
    "totalFees": 1,
    "totalPremium": 45,
    "underlying": "AAPL",
  },
]
`;

exports[`groupPositions Shows a position as closed if all legs were closed out 1`] = `
Array [
  Object {
    "closeDate": 2022-01-02T00:00:00.000Z,
    "legs": Array [
      Object {
        "closeDate": 2022-01-02T00:00:00.000Z,
        "expiration": "2022-01-01",
        "fees": 1,
        "openDate": 2022-01-01T00:00:00.000Z,
        "premium": 45,
        "quantity": 1,
        "realizedProfitLoss": 33,
        "side": "Short",
        "state": "Closed",
        "strike": 180,
        "symbol": "AAPL220101C00180000",
      },
      Object {
        "closeDate": 2022-01-02T00:00:00.000Z,
        "expiration": "2022-01-01",
        "fees": 1,
        "openDate": 2022-01-01T00:00:00.000Z,
        "premium": 45,
        "quantity": 1,
        "realizedProfitLoss": 33,
        "side": "Short",
        "state": "Closed",
        "strike": 180,
        "symbol": "AAPL220101C00180000",
      },
    ],
    "openDate": 2022-01-01T00:00:00.000Z,
    "realizedProfitLoss": 66,
    "state": "Closed",
    "totalFees": 2,
    "totalPremium": 90,
    "underlying": "AAPL",
  },
]
`;

exports[`groupPositions Shows a position as open if legs are open 1`] = `
Array [
  Object {
    "closeDate": null,
    "legs": Array [
      Object {
        "closeDate": null,
        "expiration": "2022-01-01",
        "fees": 1,
        "openDate": 2022-01-01T00:00:00.000Z,
        "premium": 45,
        "quantity": 1,
        "realizedProfitLoss": 0,
        "side": "Short",
        "state": "Open",
        "strike": 180,
        "symbol": "AAPL220101C00180000",
      },
    ],
    "openDate": 2022-01-01T00:00:00.000Z,
    "realizedProfitLoss": 0,
    "state": "Open",
    "totalFees": 1,
    "totalPremium": 45,
    "underlying": "AAPL",
  },
]
`;

exports[`groupPositions Shows a position as open if legs are open but some legs have been closed 1`] = `
Array [
  Object {
    "closeDate": null,
    "legs": Array [
      Object {
        "closeDate": 2022-01-02T00:00:00.000Z,
        "expiration": "2022-01-01",
        "fees": 1,
        "openDate": 2022-01-01T00:00:00.000Z,
        "premium": 45,
        "quantity": 1,
        "realizedProfitLoss": 33,
        "side": "Short",
        "state": "Closed",
        "strike": 180,
        "symbol": "AAPL220101C00180000",
      },
      Object {
        "closeDate": null,
        "expiration": "2022-01-01",
        "fees": 1,
        "openDate": 2022-01-01T00:00:00.000Z,
        "premium": 45,
        "quantity": 1,
        "realizedProfitLoss": 0,
        "side": "Short",
        "state": "Open",
        "strike": 180,
        "symbol": "AAPL220101C00180000",
      },
    ],
    "openDate": 2022-01-01T00:00:00.000Z,
    "realizedProfitLoss": 33,
    "state": "Open",
    "totalFees": 2,
    "totalPremium": 90,
    "underlying": "AAPL",
  },
]
`;
